### PR TITLE
add attribute & rule support

### PR DESCRIPTION
**Pull request by chylli-deriv**

### CODE REVIEW

File: dist.ini [hunk 1]
```diff
@@ -6,7 +6,5 @@ copyright_year   = 2022
 
 [@Author::DERIV]
 allow_dirty = lib/WebService/GrowthBook.pm
-[PerlTidy]
-perltidyrc = t/rc/perltidyrc
```
Comment (chylli-deriv) [inline dist.ini:10]: perltidy will complain Object::Pad's keyword. so remove it.

File: examples/feature.pl [hunk 2]
```diff
@@ -13,3 +13,7 @@
 print "string feature:", $gb->get_feature_value('string-feature'),"\n";
 print "number feature:", $gb->get_feature_value('number-feature'),"\n";
 print "json feature:", Dumper($gb->get_feature_value('json-feature')),"\n";
+$gb->set_attributes({ id => 1});
+print "bool attribute: ", ($gb->is_on('bool-attribute')),"\n";
+$gb->set_attributes({ id => 123});
+print "bool attribute: ", ($gb->is_on('bool-attribute')),"\n";
```
Comment (chylli-deriv) [inline examples/feature.pl]:
```suggestion
print "bool attribute: ", ($gb->is_on('bool-attribute')),"\n";

```

File: lib/WebService/GrowthBook.pm [hunk 3]
```diff
@@ -63,45 +96,750 @@ class WebService::GrowthBook {
         $features = {};
         for my $feature_id (keys $features_set->%*) {
             my $feature = $features_set->{$feature_id};
+            use Data::Dumper;
```
Comment (chylli-deriv) [inline lib/WebService/GrowthBook.pm]:
```suggestion
```

File: lib/WebService/GrowthBook.pm [hunk 4]
```diff
@@ -67,41 +100,745 @@ class WebService::GrowthBook {
...
+
     method is_off($feature_name) {
         my $result = $self->eval_feature($feature_name);
         return undef unless defined($result);
         return $result->off;
     }
-    
-    method eval_feature($feature_name){
+
+    # I don't know why it is called stack. In fact it is a hash/dict
```
Comment (chylli-deriv) [inline lib/WebService/GrowthBook.pm]:
```suggestion
    # I don't know why it is called stack in python version SDK. In fact it is a hash/dict
```

File: lib/WebService/GrowthBook/Eval.pm [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+    if (ref($attribute_value) eq 'HASH') {
+        return "object";
+    }
+    if (ref($attribute_value) eq 'SCALAR' && ($$attribute_value eq '0' || $$attribute_value eq '1')) {
+        return "boolean";
+    }
+    return "unknown";
+}
+
+1;
```
Comment (chylli-deriv) [inline lib/WebService/GrowthBook/Eval.pm]:
```suggestion
1;

```

File: lib/WebService/GrowthBook/FeatureResult.pm [hunk 6]
```diff
@@ -8,19 +8,34 @@ use JSON::MaybeXS;
 
 ## VERSION
 
+# TODO change id to feature_id
```
Comment (chylli-deriv) [inline lib/WebService/GrowthBook/FeatureResult.pm]:
```suggestion
```

File: lib/WebService/GrowthBook/Experiment.pm [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        $obj{bucket_version} = bucket_version if $bucket_version;
+        $obj{min_bucket_version} = min_bucket_version if $min_bucket_version;
+        $obj{parent_conditions} = parent_conditions if $parent_conditions;
+
+        return \%obj;
+    }
+}
+
+
+1;
```
Comment (chylli-deriv) [inline lib/WebService/GrowthBook/Experiment.pm]:
```suggestion
1;

```

File: lib/WebService/GrowthBook/Result.pm [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        $obj{name} = $name if $name;
+        $obj{passthrough} = 1 if $self->passthrough;
+
+        return \%obj;
+    }
+
+
+}
+
+1;
```
Comment (chylli-deriv) [inline lib/WebService/GrowthBook/Result.pm]:
```suggestion
1;

```

File: t/growthbook.t [hunk 9]
```diff
@@ -37,9 +36,10 @@ is($instance->get_feature_value('string-feature'), 'A string', 'string-feature v
...
-is($instance->is_off('not-exist-feature'), undef, "not-exist-feature is undef");
-is($instance->is_on('string-feature'), undef, "is_on string is undef");
-is($instance->is_off('string-feature'), undef, "is_off string is undef");
+is($instance->get_feature_value('not-exist-feature', 123), 123, "fallback value");
+is($instance->is_on('not-exist-feature'), 0, "not-exist-feature is undef");
+is($instance->is_off('not-exist-feature'), 1, "not-exist-feature is undef");
+is($instance->is_on('string-feature'), 1, "is_on string is true");
+is($instance->is_off('string-feature'), 0, "is_off string is false");
 
 done_testing();
```
Comment (chylli-deriv) [inline t/growthbook.t]:
```suggestion
done_testing();

```

File: lib/WebService/GrowthBook.pm [hunk 10]
```diff
@@ -45,11 +50,39 @@ WebService::GrowthBook - sdk of growthbook
...
+
     field $cache //= WebService::GrowthBook::InMemoryFeatureCache->singleton;
+    field $sticky_bucket_assignment_docs //= {};
+    field $tracked = {};
+    field $assigned = {};
+    field $subscriptions = [];
+
+    sub BUILDARGS{
+        my ($class, %args) = @_;
+        adjust_args_camel_to_snake(\%args);
```
Comment (nihal-deriv) [inline lib/WebService/GrowthBook.pm:76]: just curios why do we convert camel_case arguments to snake_case ?

Comment (chylli-deriv) [reply lib/WebService/GrowthBook.pm:76]: because the data loaded from growthbook json file is camel_case, but the perl style is snake_case.

**Pull request merged**